### PR TITLE
Race condition and test fixes

### PR DESCRIPTION
**Pull request by mvdan**
Picked up from @mikedanese's #12967

The reflect test failure fix really looks like a Go bug. I'll try and find any reference to it.

### CODE REVIEW

**Comment by mvdan** [comment]
@mikedanese could you confirm you're okay with this so that `cla: no` can go away?

**Comment by mikedanese** [comment]
I'm OK with these contributions. @vishh for cadvisor upgrade, @thockin or @brendandburns for the go upgrade.

**Comment by mikedanese** [comment]
@k8s-bot ok to test.

**Comment by mvdan** [comment]
`TestHTTPProbeChecker` appears to be failing now, which didn't fail before. Will look into it.

**Comment by vishh** [comment]
LGTM for cadvisor changes.

**Comment by mvdan** [comment]
Currently talking to a member of the Go team about the weird if condition possible bug.

**Comment by mvdan** [comment]
Confirmed - it's a bug in the compiler. My fix works around the issue, so better keep it in to keep the code working in 1.5.

**Comment by thockin** [comment]
Achievement Unlocked?

I'm happy to see us support 1.5

**Comment by mvdan** [comment]
@thockin double achievement unlocked - 1.5 support and compiler bug spotted :)

For anyone interested, the compiler bug is https://github.com/golang/go/issues/12266

**Comment by mvdan** [comment]
Fixed the `TestHTTPProbeChecker` failure in 1.5, but that will make it fail on 1.4 and earlier. Is that acceptable? Or should we not check GET bodies on `probe.Failure` at all?

**Comment by smarterclayton** [comment]
We have to maintain support for building and testing on 1.4 for a while at least.

**Comment by thockin** [comment]
of course

**Comment by smarterclayton** [comment]
Was responding to

>  but that will make it fail on 1.4 and earlier. Is that acceptable?

**Comment by thockin** [comment]
oh, right.  No, 1.4 has to work.

**Comment by mvdan** [comment]
Sure. Amended the last commit to support both the old and new error messages.

Still a few failing tests to go.

**Comment by mvdan** [comment]
None of the tests still failing in shippable fail on my machine. Anyone know why isn't shippable running now?

**Comment by mvdan** [comment]
Uncovered a bad race condition bug while investigating the portforward test failure on 1.5. It is now fixed, the test passes consistently on my machine now.

I wonder how it could have gone unnoticed in 1.4. Maybe the runtime being slower hid the race condition?

**Comment by mvdan** [comment]
I found another flaky test:

```
--- FAIL: TestRequestAttachRemoteCommand (30.01s)
        remotecommand_test.go:241: 0: expected error stream read 'bail', got 'Timeout occured'
```

This one I can't reproduce, so I'm ignoring it for now. Or should I open an issue about it?

**Comment by mvdan** [comment]
In fact, shippable had errors in remotecommand too. Although the errors are different, it probably means that there's another underlying bug that 1.5 is uncovering. Will investigate.

**Comment by mvdan** [comment]
Fixed too. Let's see if shippable passes now.

Any chance we could have shippable run five times in a row, to be sure there aren't any more race condition bugs popping up thanks to 1.5?

**Comment by mvdan** [comment]
All the tests are passing, but `hack/test-cmd.sh` is failing due to a template issue.

Seems like the template package used to allow `{{..foo}}` instead of `{{.foo}}`, and 1.5 simply doesn't. Will try a fix.

**Comment by mvdan** [comment]
Success! There's only one integration test failing now. Will look at it tomorrow.

**Comment by mvdan** [comment]
Shippable uncovers more and more flaky tests due to race conditions at every run, so don't know when I'll be done. Merging in 1.5 before these are fixed would be unwise, since apparently they aren't flaky when running on 1.4.

**Comment by mvdan** [comment]
It just occured to me that all the new flaky tests are showing up because the scheduler in 1.4 is deterministic, but the one in 1.5 is randomized.

All the bugs I'm fixing now aren't present in 1.4, because we were lucky that the deterministic scheduling didn't trigger the race conditions.

**Comment by mvdan** [comment]
Shippable and jenkins have passed three times in a row, so I would say this is ready.

**Comment by mvdan** [comment]
Any reviewers?

**Comment by mvdan** [comment]
Rebased on master, tests still pass on my machine.

**Comment by mvdan** [comment]
Oh, go's `-race` found another bug. Will fix.

File: pkg/client/unversioned/remotecommand/remotecommand_test.go [hunk 1]
```diff
@@ -92,15 +92,21 @@ func fakeExecServer(t *testing.T, i int, stdinData, stdoutData, stderrData, erro
...
 		if len(stdoutData) > 0 {
 			fmt.Fprint(stdoutStream, stdoutData)
+		}
+		if stdoutStream != nil {
 			stdoutStream.Close()
 		}
 		if len(stderrData) > 0 {
 			fmt.Fprint(stderrStream, stderrData)
+		}
+		if stderrStream != nil {
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand_test.go]: I touched this in https://github.com/kubernetes/kubernetes/pull/13144

It's actually better to avoid calling Close() on stdoutStream/stderrStream and let the deferred Reset() calls close the streams. Calling Close() explicitly actually introduces a race with the stdinStream read below

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 2]
```diff
@@ -155,18 +156,16 @@ func (e *Streamer) doStream() error {
 	}
 	defer conn.Close()
 
-	doneChan := make(chan struct{}, 2)
 	errorChan := make(chan error)
+	var wg sync.WaitGroup
 
 	cp := func(s string, dst io.Writer, src io.Reader) {
+		defer wg.Done()
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: this gets called for stdin, which won't terminate until we tear it down

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 3]
```diff
@@ -195,49 +195,37 @@ func (e *Streamer) doStream() error {
...
-		select {
-		case <-doneChan:
-			completedStreams++
-			if completedStreams == waitCount {
-				break Loop
-			}
-		case err := <-errorChan:
-			return err
-		}
+	wg.Wait()
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: waiting here means a hung stdout/stderr stream can keep us from returning, even if we encountered an error

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 4]
```diff
@@ -155,18 +156,16 @@ func (e *Streamer) doStream() error {
 	}
 	defer conn.Close()
 
-	doneChan := make(chan struct{}, 2)
 	errorChan := make(chan error)
+	var wg sync.WaitGroup
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: cc @ncdc

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 3 above)
Comment (mvdan) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: That was the original intended behaviour. What do you suggest instead of waiting?

File: pkg/client/unversioned/remotecommand/remotecommand_test.go (hunk 1 above)
Comment (mvdan) [inline pkg/client/unversioned/remotecommand/remotecommand_test.go]: If I remove the Close() calls, the tests just hang. The commit message explains why - if we don't close the error stream, another goroutine blocks forever waiting for more errors.

Does your PR fix that?

File: pkg/client/unversioned/remotecommand/remotecommand.go [hunk 5]
```diff
@@ -195,49 +195,37 @@ func (e *Streamer) doStream() error {
 			return err
 		}
 		defer remoteStdin.Reset()
-		// TODO this goroutine will never exit cleanly (the io.Copy never unblocks)
-		// because stdin is not closed until the process exits. If we try to call
-		// stdin.Close(), it returns no error but doesn't unblock the copy. It will
-		// exit when the process exits, instead.
+		wg.Add(1)
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: @ncdc will waiting stdin to close itself hang here?

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 2 above)
Comment (mvdan) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: I did not run into that stdin issue with the unit tests, see the two commits in this PR touching this file.

Or does that problem show up somewhere other than unit/integration tests?

File: pkg/client/unversioned/remotecommand/remotecommand_test.go (hunk 1 above)
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand_test.go]: remove Close() and switch the stdin ioutil.ReadFull to io.ReadAll(expectedLen)

Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand_test.go]: This test shouldn't have to close anything explicitly... all the streams get Reset() and conn.Close() is called when the function returns. The ioutil.ReadFull() below was what was hanging

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 2 above)
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: The unit tests mock reality... @ncdc can probably give more details about the behavior of stdin streams in real life

File: pkg/client/unversioned/remotecommand/remotecommand_test.go (hunk 1 above)
Comment (mvdan) [inline pkg/client/unversioned/remotecommand/remotecommand_test.go]: I see, trying that out now.

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 2 above)
Comment (mvdan) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: I just knew that not waiting for stdin to finish reading is definitely wrong, so it seemed to me to better try and trigger the bug than to leave in code that I knew was wrong. And I just failed at reproducing the bug, so I left that in :)

File: pkg/client/unversioned/remotecommand/remotecommand.go (hunk 3 above)
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/remotecommand.go]: more noting the change than objecting to it... I like waiting for all three streams, but I wasn't sure if there was a situation where an error propagated via the errorStream was expected to exit early and save us from a hung stderr/stdout stream

**Comment by mvdan** [comment]
Another race condition fixed.

There's an e2e test that keeps on failing, anyone got any ideas? This should speak to someone:

`Unexpected number of endpoints: found map[92ef79f7-4b58-11e5-90f1-42010af0bcc6:[100]], expected map[podname1:[100] podname2:[101]]`

Part of this test:

```
[It] should serve multiport endpoints from pods
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:236
```

**Comment by mvdan** [comment]
@liggitt with the stdin changes reverted, remotecommands is flaky again. I'll keep on digging...

**Comment by liggitt** [comment]
Do you have a copy of the flake?

**Comment by mvdan** [comment]
@liggitt here you go:

```
--- FAIL: TestRequestExecuteRemoteCommand (0.01s)
        remotecommand_test.go:184: 1: expected stderr data 'c', got ''
```

Although, surprisingly enough, making it wait for stdin to finish copying doesn't fix it. Currently running the test 100 times to check for flakyness.

**Comment by liggitt** [comment]
Yeah, the readfull prior to closing any of the streams that execute is waiting for should have covered it :-/

**Comment by mvdan** [comment]
It doesn't appear to be a race condition - even a `time.Sleep(100 * time.Millisecond)` before the `wg.Wait()` doesn't fix it.

**Comment by mvdan** [comment]
@liggitt found it - it's the fake exec server that is racey. Making it only handle one HTTP connection at a time via a mutex fixes the issue. This one was hard to find...

**Comment by mvdan** [comment]
The race detected in the oidc package isn't actually our fault: https://github.com/golang/go/issues/12262

**Comment by mvdan** [comment]
To summarize:

* There is a flake in remotecommand (see comments above) but it's not introduced by Go 1.5, as @liggitt mentioned it happens on 1.4 too
* There is a flaky e2e test: `[It] should serve multiport endpoints from pods` repeatedly fails with `Unexpected number of endpoints` until timeout is reached. Has anyone seen it before? Any ideas?

**Comment by smarterclayton** [comment]
Services e2e tests tend to be flaky because they are dependent on
scheduling several pods and wait a short time period while parallel tests
are running.  I don't believe we have a real flake there, but we haven't
proved it yet.

On Aug 26, 2015, at 3:11 AM, Daniel Martí <notifications@github.com> wrote:

To summarize:

   - There is a flake in remotecommand (see comments above) but it's not
   introduced by Go 1.5, as @liggitt <https://github.com/liggitt> mentioned
   it happens on 1.4 too
   - There is a flaky e2e test: [It] should serve multiport endpoints from
   pods repeatedly fails with Unexpected number of endpoints until timeout
   is reached. Has anyone seen it before? Any ideas?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13058#issuecomment-134850249>
.

**Comment by mvdan** [comment]
So since as far as I can tell this PR doesn't add any flakyness, I think it's ready to be merged.

Ping @mikedanese @thockin

**Comment by thockin** [comment]
@mikedanese for final

**Comment by ncdc** [comment]
I just noticed there are port forwarding changes. I need some time to review as I'm afraid they might cause leaked goroutines.

**Comment by ncdc** [comment]
I have a PR that I'm working on to address some port forward issues (in relation to #8766) but I'm spread a bit thin at the moment and haven't had time to devote to it recently.

**Comment by ncdc** [comment]
Actually, if the tests pass, this works for me. I'm working on some new e2e tests that will test various port forwarding scenarios to make sure there aren't any hangs or leaks.

**Comment by mvdan** [comment]
@ncdc thanks for the info. I only did the minimal to fix the race conditions and get the tests to consistently pass - I highly doubt that introduced any bugs, especially in portforward where the changes are minimal.

**Comment by ncdc** [comment]
@mvdan my fear is this won't ever be called due to the root cause of 8766: https://github.com/kubernetes/kubernetes/pull/13058/files#diff-3c23f834eaf7e756813f19c5a1116c46R295

**Comment by mvdan** [comment]
@ncdc so you're saying that currently two bugs cancel each other out, and that if I fix this one the other will make the `Wait()` call to block forever?

**Comment by mikedanese** [comment]
@mvdan thanks for fixing all these races. I will try to review shortly. However I am hesitant to merge the go1.5 change. I dislike that we had to restructure correct if statements to make go1.5 work. Doesn't it seem like we should wait for go1.5.1 to fix some of the multiple regressions we ran into here before bumping the version we run in CI?

**Comment by mvdan** [comment]
@mikedanese following the same logic, why are we using 1.4 instead of 1.4.2?

1.5.1 is scheduled for mid-September. Waiting for that to bump the CI version seems fine to me - I can remove your commit from this PR - but the rest of commits should still be merged.

@ncdc and myself have come to a solution regarding portforward, I'll amend the commit in a second.

**Comment by liggitt** [comment]
+1 on merging race fixes

**Comment by mikedanese** [comment]
We didn't have to restructure correct if statements when we moved to 1.4 so the same logic does not apply. I agree that we should merge race fixes. I only said "I am hesitant to merge the go1.5 changes".

**Comment by mvdan** [comment]
@mikedanese little bugs like these got fixed between 1.4 and 1.4.2, surely some affect kubernetes and weren't caught in unit/e2e tests. I agree that this bug is probably bigger than any other, but I would always use the latest stable release of each major Go release regardless of whether it fixes any regressions/bugs or not.

Just to confirm - by go1.5 changes, you mean the shippable changes and any changes that work around bugs in 1.5, right?

**Comment by mvdan** [comment]
In fact, the only go 1.5 workaround there is the if change, which doesn't break 1.4. If we don't merge it, kubernetes will remain broken on the current Go release until 1.5.1 is out. I think that's worse.

**Comment by mvdan** [comment]
I have removed the portforward race fix since it requires server side fixes too, and it isn't introduced by 1.5. @ncdc is taking care of it in https://github.com/kubernetes/kubernetes/pull/12283.

**Comment by mvdan** [comment]
Also removed the remotecommand race fix since it's been cherry picked by @liggitt in #13144

**Comment by timothysc** [comment]
@wojtek-t @fgrzadkowski very interested to see the gc-perf difference on 1.5 for the api-server.

**Comment by mikedanese** [comment]
@mvdan I am referring to only the shippable.yml and .travis.yml changes when I refer to go1.5 changes. All the bug fixes are definitely worthwhile. I will review this tomorrow or this weekend. If other people feel strongly that we should bump ci to go1.5 I won't veto, but  in light of the nasty if bug I tend against it.

**Comment by mvdan** [comment]
@mikedanese agreed. I'll remove your commit from this PR.

File: third_party/forked/reflect/deep_equal.go [hunk 6]
```diff
@@ -156,7 +156,9 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
 		}
 		return true
 	case reflect.Slice:
-		if (v1.IsNil() || v1.Len() == 0) != (v2.IsNil() || v2.Len() == 0) {
+		empty1 := v1.IsNil() || v1.Len() == 0
```
Comment (mikedanese) [inline third_party/forked/reflect/deep_equal.go]: can you leave a comment here that explains why this change was necessary?

File: pkg/probe/http/http_test.go [hunk 7]
```diff
@@ -39,14 +48,31 @@ func TestHTTPProbeChecker(t *testing.T) {
 
 	prober := New()
 	testCases := []struct {
-		handler func(w http.ResponseWriter)
-		health  probe.Result
-		body    string
+		handler   func(w http.ResponseWriter)
+		health    probe.Result
+		accBodies []string
```
Comment (mikedanese) [inline pkg/probe/http/http_test.go]: can you leave a comment here that explains why this change was necessary? (potentially include something I could grep for like `// go1.5: the error message changed`)

File: third_party/forked/reflect/deep_equal.go (hunk 6 above)
Comment (mvdan) [inline third_party/forked/reflect/deep_equal.go]: sure.

File: pkg/probe/http/http_test.go (hunk 7 above)
Comment (mvdan) [inline pkg/probe/http/http_test.go]: sure.

File: pkg/client/unversioned/record/event_test.go [hunk 8]
```diff
@@ -316,17 +319,17 @@ func validateEvent(actualEvent *api.Event, expectedEvent *api.Event, t *testing.
 	if actualEvent.FirstTimestamp.IsZero() || actualEvent.LastTimestamp.IsZero() {
 		t.Errorf("timestamp wasn't set: %#v", *actualEvent)
 	}
-	if actualEvent.FirstTimestamp.Equal(actualEvent.LastTimestamp) {
+	actualFirstTimestamp := actualEvent.FirstTimestamp
```
Comment (mikedanese) [inline pkg/client/unversioned/record/event_test.go]: go1.5 comment please

Comment (mvdan) [inline pkg/client/unversioned/record/event_test.go]: This isn't a go1.5 workaround/fix. Just a small code simplification I did along the way while I fixed the race bug.

File: pkg/client/unversioned/record/event_test.go [hunk 9]
```diff
@@ -286,24 +289,24 @@ func TestEventf(t *testing.T) {
...
 				}
-				called <- struct{}{}
+				wg.Done()
 				return returnEvent, nil
 			},
 		}
 		eventBroadcaster := NewBroadcaster()
 		sinkWatcher := eventBroadcaster.StartRecordingToSink(&testEvents)
 		logWatcher1 := eventBroadcaster.StartLogging(t.Logf) // Prove that it is useful
+		wg.Add(1)
```
Comment (mikedanese) [inline pkg/client/unversioned/record/event_test.go]: Instead of adding a wait group, why not just move one of the `<-called` to this line? Isn't that the same?

File: pkg/client/unversioned/record/event_test.go (hunk 8 above)
Comment (mikedanese) [inline pkg/client/unversioned/record/event_test.go]: Ah ok.

File: pkg/client/unversioned/record/event_test.go (hunk 9 above)
Comment (mvdan) [inline pkg/client/unversioned/record/event_test.go]: A waitgroup is simpler to use and better (e.g. it lets goroutines die sooner).

The original code did the waiting right before the Stop() calls, so I did the same with a waitgroup. Splitting the waiting in two separate blocks of code would confuse things, especially since you cannot know which of the two channel receives would correspond to which of the channel sends.

**Comment by mikedanese** [comment]
https://github.com/mvdan/kubernetes/commit/2527da99e1f826febc0cd97b0d3b42a1713f08b6 looks fine to me but cc @karlkfi @sttts @jdef for a second review on that commit.

**Comment by mikedanese** [comment]
Generally LGTM. I would like @lavalamp to review https://github.com/mvdan/kubernetes/commit/579118752230211aca5eb23e74dc4ccfc6457590  before final approval.

**Comment by mvdan** [comment]
FWIW, after the rebase the two commits that @mikedanese linked to are now https://github.com/mvdan/kubernetes/commit/d3490e86ec34abd785d0a1974b09738149ce71ef and https://github.com/mvdan/kubernetes/commit/2ab6033cb0b94c3321796893e34744fedc9a9a46, respectively.

File: third_party/forked/reflect/deep_equal.go [hunk 10]
```diff
@@ -156,7 +156,10 @@ func (e Equalities) deepValueEqual(v1, v2 reflect.Value, visited map[visit]bool,
 		}
 		return true
 	case reflect.Slice:
-		if (v1.IsNil() || v1.Len() == 0) != (v2.IsNil() || v2.Len() == 0) {
+		// go1.5: simplified if to work around https://github.com/golang/go/issues/12266
```
Comment (lavalamp) [inline third_party/forked/reflect/deep_equal.go]: Let's not have these workarounds in our code, please--let's wait for a go1.5 patch before using go1.5

File: pkg/watch/mux.go [hunk 11]
```diff
@@ -149,6 +151,7 @@ func (m *Broadcaster) Action(action EventType, obj runtime.Object) {
 // You must not call Action after calling Shutdown.
```
Comment (lavalamp) [inline pkg/watch/mux.go]: Action or Watch

File: third_party/forked/reflect/deep_equal.go (hunk 10 above)
Comment (mvdan) [inline third_party/forked/reflect/deep_equal.go]: You can argue that they should be removed once the current version of Go is 1.5.1. But for as long as the stable is 1.5, not including this patch will mean that you can't use Kubernetes with the current stable version, not even for testing/development purposes.

File: pkg/watch/mux.go (hunk 11 above)
Comment (lavalamp) [inline pkg/watch/mux.go]: Also, say that it blocks until all events have been distributed to all extant watchers.

Comment (mvdan) [inline pkg/watch/mux.go]: Sure.

File: pkg/watch/mux.go [hunk 12]
```diff
@@ -162,6 +165,7 @@ func (m *Broadcaster) loop() {
 		}
 		m.distribute(event)
 	}
+	m.distributing.Done()
```
Comment (lavalamp) [inline pkg/watch/mux.go]: Do this after the closeAll()?

File: pkg/watch/mux.go (hunk 11 above)
Comment (lavalamp) [inline pkg/watch/mux.go]: Clarification: distributed to, but not necessarily handled by, all extant watchers; stuff could remain sitting in the outbound channels.

File: pkg/watch/mux.go (hunk 12 above)
Comment (mvdan) [inline pkg/watch/mux.go]: Which would result in closing the channels before we waited for all the info to have been sent to them. I don't see why.

File: pkg/controller/framework/controller_test.go [hunk 13]
```diff
@@ -371,11 +366,18 @@ func TestUpdate(t *testing.T) {
 		},
 	}
 
-	// run every test a few times, in parallel
 	const threads = 3
+	testDoneWG.Add(threads * len(tests))
```
Comment (lavalamp) [inline pkg/controller/framework/controller_test.go]: Nit: slight preference for moving this line up instead of moving the controller start down here.

File: pkg/watch/mux.go (hunk 11 above)
Comment (mvdan) [inline pkg/watch/mux.go]: You mean blocked channel sends because the receivers aren't listening yet? In that case, since our channel send has blocked waiting for a listener, the Done() call would not be done and we would wait. So all events must have been sent, nothing can remain sitting in outbound channels - at least with the current approach.

File: pkg/client/unversioned/record/event_test.go [hunk 14]
```diff
@@ -271,14 +272,16 @@ func TestEventf(t *testing.T) {
...
+		// We expect only one callback
+		wg.Add(1)
 		testEvents := testEventSink{
 			OnCreate: func(event *api.Event) (*api.Event, error) {
 				returnEvent, _ := validateEvent(event, item.expect, t)
 				if item.expectUpdate {
 					t.Errorf("Expected event update(), got event create()")
 				}
-				called <- struct{}{}
+				wg.Done()
```
Comment (lavalamp) [inline pkg/client/unversioned/record/event_test.go]: Use `defer wg.Done()` at the beginning of the callback.

File: pkg/controller/framework/controller_test.go (hunk 13 above)
Comment (mvdan) [inline pkg/controller/framework/controller_test.go]: Sure. My intention was to keep `const threads = 3` close to the code that uses it below, but what you suggest would mean fewer line changes so it makes sense.

File: pkg/client/unversioned/record/event_test.go (hunk 14 above)
Comment (mvdan) [inline pkg/client/unversioned/record/event_test.go]: Nice catch. Although, to my defense, this error was already there :)

File: pkg/client/unversioned/record/event_test.go (hunk 9 above)
Comment (lavalamp) [inline pkg/client/unversioned/record/event_test.go]: Why not start the wait group off with 2 to begin with?

Comment (mvdan) [inline pkg/client/unversioned/record/event_test.go]: The idea is to keep each `Add(1)` close to what will later trigger the `Done()` call. If you do a single `Add(n)` at the top, it might later be harder to spot what blocks of code amount for how many of the `Done()` calls.

This of course doesn't apply to loops.

Comment (liggitt) [inline pkg/client/unversioned/record/event_test.go]: +1 for pairing Add/Done calls

File: pkg/watch/mux.go (hunk 12 above)
Comment (lavalamp) [inline pkg/watch/mux.go]: No-- a channel close doesn't blow away all the data in the channel, it just prevents further writes.

File: pkg/watch/mux.go (hunk 11 above)
Comment (lavalamp) [inline pkg/watch/mux.go]: Outbound channels could be buffered, and even if they weren't, you don't wait for close to have been called on them.

File: pkg/watch/mux.go (hunk 12 above)
Comment (mvdan) [inline pkg/watch/mux.go]: The writes may not even have started, that's my point.

File: pkg/watch/mux.go (hunk 11 above)
Comment (mvdan) [inline pkg/watch/mux.go]: I had not noticed that our channels could be buffered, you're right.

I can't find any reference to this, but I am guessing that if you close a buffered channel, it can still be read from and its buffered values are not lost.

File: contrib/mesos/pkg/proc/proc.go [hunk 15]
```diff
@@ -122,11 +121,13 @@ func (self *procImpl) begin() runtime.Signal {
 	var entered runtime.Latch
 
 	// execute actions on the backlog chan
+	self.wg.Add(1)
```
Comment (jdef) [inline contrib/mesos/pkg/proc/proc.go]: this package has been completely re-implemented, see #10457 (lgtm'd, awaiting merge)

can we avoid making changes here for now?

Comment (mvdan) [inline contrib/mesos/pkg/proc/proc.go]: Sure.

File: pkg/watch/mux.go (hunk 11 above)
Comment (lavalamp) [inline pkg/watch/mux.go]: >  if you close a buffered channel, it can still be read from and its buffered values are not lost.

Yes, that's correct.

File: pkg/watch/mux.go (hunk 12 above)
Comment (lavalamp) [inline pkg/watch/mux.go]: Distribute is called synchronously above. Once we get to this point, no more writes to the outgoing channels will occur, unless you count the close call, which I do. After closeAll is called, then there really will not be any more writes to the channels and it should be safe to tell the waitgroup that we're done.

Comment (mvdan) [inline pkg/watch/mux.go]: You are right, the channel sends are synchronous and cannot happen at this point. And they must all have been received by this point too, so even if closing a channel cancels blocked unbuffered channel sends - which I'm not sure about - this would never happen.

Nice catch!

**Comment by mvdan** [comment]
Thanks for the throrough code review @lavalamp! Much appreciated.

**Comment by lavalamp** [comment]
Sure, no problem, thanks for clean up!

**Comment by mvdan** [comment]
All comments amended.

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by gmarek** [comment]
cc @gmarek @fgrzadkowski @wojtek-t

**Comment by mvdan** [comment]
Go 1.5.1 was just released: https://github.com/golang/go/issues?q=milestone%3AGo1.5.1

@mikedanese enabling that in the CI could now be done.

**Comment by mikedanese** [comment]
@mvdan please do reenable. LGTM when CI is green

**Comment by mvdan** [comment]
`gimme` doesn't know about 1.5.1 yet, it was released just hours ago. I'll look again tomorrow.

**Comment by mikedanese** [comment]
@mvdan sounds good. for context, I was trying to switch shippable to gimme, because IIUC gvm builds go from source and post 1.5 building from source requires a compiled go to bootstrap the compiler. Thus we would have to build multiple versions of go per shippable run

**Comment by mvdan** [comment]
@mikedanese in fact, let's keep the 1.5.1 bump separate. You might want to build kubernetes with 1.5.1 instead of 1.4 for official builds too.

Shippable failed, but I keep on getting 500's on all of its builds. This PR should not be adding any failures anyway. If it does, or if there are any more issues arising thanks to 1.5, feel free to ping me.

**Comment by mikedanese** [comment]
Shippable says all tests pass but
```
+++ [0909 17:39:00] Saved JUnit XML test report to /root/workspace/src/github.com/kubernetes/kubernetes/shippable/testresults/junit_echo_20150909-172222.xml
!!! Error in ./hack/test-go.sh:210
  'return ${rc}' exited with status 2
Call stack:
  1: ./hack/test-go.sh:210 main(...)
Exiting with status 1
```
@ixdy do you know what's going on here? It's not readily obvious.

**Comment by mvdan** [comment]
@mikedanese shippable was acting up yesterday, maybe it will work properly now.

Removed the Go 1.5 bug workaround, since 1.5.1 is now out.

**Comment by mikedanese** [comment]
Fixes #13764

File: pkg/controller/framework/controller_test.go [hunk 16]
```diff
@@ -335,7 +335,12 @@ func TestUpdate(t *testing.T) {
 		},
 	)
 
+	const threads = 3
+	testDoneWG.Add(threads * len(tests))
```
Comment (ixdy) [inline pkg/controller/framework/controller_test.go]: The issue with Shippable is that this test is failing to build, probably because `tests` isn't defined yet when it's used here.

Comment (mvdan) [inline pkg/controller/framework/controller_test.go]: You're right. Whoops.

This used to compile, but when @lavalamp suggested to reorganize the code I didn't see this error. The only sane way to fix this is to move the code back where it was, effectively undoing the suggestion from Daniel.

Comment (lavalamp) [inline pkg/controller/framework/controller_test.go]: Oops.
